src/merge.rs

```diff
@@ -192,3 +194,295 @@ impl<'repo> Drop for AnnotatedCommit<'repo> {
+        self
+    }
+
+    /// Specify marker size, default is 7: <<<<<<< ours
+    pub fn marker_size(&mut self, size: u16) -> &mut MergeFileOptions {
+        self.raw.marker_size = size;
+        self
+    }
+
+    /// Acquire a pointer to the underlying raw options.
+    pub unsafe fn raw(&self) -> *const raw::git_merge_file_options {
+        &self.raw as *const _
+    }
+}
+
+/// For git_merge_file_input
+pub struct MergeFileInput {
+    raw: raw::git_merge_file_input,
+
+    /// File name of the conflicted file, or `NULL` to not merge the path.
+    ///
+    /// You can turn this value into a `std::ffi::CString` with
+    /// `CString::new(&entry.path[..]).unwrap()`. To turn a reference into a
+    /// `&std::path::Path`, see the `bytes2path()` function in the private,
+    /// internal `util` module in this crate’s source code.
+    path: Option<CString>,
+
+    /// File mode of the conflicted file, or `0` to not merge the mode.
+    pub mode: Option<FileMode>,
+
+    /// File content
+    pub content: Option<Vec<u8>>,
+}
+
+impl Default for MergeFileInput {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MergeFileInput {
+    /// Creates a new set of empty diff options.
+    pub fn new() -> MergeFileInput {
+        let mut input = MergeFileInput {
+            raw: unsafe { mem::zeroed() },
+            path: None,
+            mode: None,
+            content: None,
+        };
+        assert_eq!(
+            unsafe { raw::git_merge_file_input_init(&mut input.raw, 1) },
+            0
+        );
+        input
+    }
+
+    /// File name of the conflicted file, or `None` to not merge the path.
+    pub fn path<T: IntoCString>(&mut self, t: T) -> &mut MergeFileInput {
+        self.path = Some(t.into_c_string().unwrap());
+
+        self.raw.path = self
+            .path
+            .as_ref()
+            .map(|s| s.as_ptr())
+            .unwrap_or(ptr::null());
+
+        self
+    }
+
+    /// File mode of the conflicted file, or `0` to not merge the mode.
+    pub fn mode(&mut self, mode: Option<FileMode>) -> &mut MergeFileInput {
+        self.mode = mode;
+
+        if let Some(mode) = self.mode {
+            self.raw.mode = mode as u32;
+        }
+
+        self
+    }
+
+    /// File content, text or binary
+    pub fn content(&mut self, content: Option<Vec<u8>>) -> &mut MergeFileInput {
+        self.content = content;
+
+        self.raw.size = self.content.as_ref().map(|c| c.len()).unwrap_or(0);
+        self.raw.ptr = self
+            .content
+            .as_ref()
+            .map(|c| c.as_ptr() as *const _)
+            .unwrap_or(ptr::null());
+
+        self
+    }
+
+    /// Get the raw struct in C
+    pub fn raw(&self) -> *const raw::git_merge_file_input {
+        &self.raw as *const _
+    }
+}
+
+impl MergeFileInput {
+    /// Create from Repository and IndexEntry
+    pub fn from(repo: &Repository, index_entry: &IndexEntry) -> MergeFileInput {
```

Comment:
FWIW I've personally shied away from helper functions like this in the past for this crate. This crate tries to stick pretty close to libgit2's own API without adding too many Rust helpers and such. Is there a particular reason though that we should have this when libgit2 doesn't?

Reply:
Understood, it's close now:
```rust
impl MergeFileInput {
    /// Create MergeFileInput
    pub fn from(path: &str, mode: FileMode, content: Vec<u8>) -> MergeFileInput {
        let mut input = MergeFileInput::new();
        input.path(path);
        input.mode(Some(mode));
        input.content(Some(content));
        input
    }
}
```